spec-cordova/prepare.spec.js

```diff
@@ -17,175 +17,265 @@
                         www: path.join(project_dir, 'platforms', platform, 'www')
                     }
                 }),
+                removePlugin: platform_api_add_mock,
+                addPlugin: platform_api_remove_mock
             };
         });
-        is_cordova = spyOn(util, 'isCordova').and.returnValue(project_dir);
-        cd_project_root = spyOn(util, 'cdProjectRoot').and.returnValue(project_dir);
-        list_platforms = spyOn(util, 'listPlatforms').and.returnValue(supported_platforms);
-        fire = spyOn(HooksRunner.prototype, 'fire').and.returnValue(Q());
-
-        find_plugins = spyOn(util, 'findPlugins').and.returnValue([]);
-        spyOn(PlatformJson, 'load').and.returnValue(new PlatformJson(null, null, {}));
-        spyOn(PlatformJson.prototype, 'save');
-        load = spyOn(lazy_load, 'based_on_config').and.returnValue(Q());
-        cp = spyOn(shell, 'cp').and.returnValue(true);
-        mkdir = spyOn(shell, 'mkdir');
-        spyOn(ConfigParser.prototype, 'write');
-        spyOn(xmlHelpers, 'parseElementtreeSync').and.callFake(function() {
-            return new et.ElementTree(et.XML(TEST_XML));
-        });
+        spyOn(PlatformJson, 'load');
+        spyOn(HooksRunner.prototype, 'fire').and.returnValue(Q());
+        spyOn(util, 'isCordova').and.returnValue(true);
     });
 
-    describe('failure', function() {
-        it('Test 001 : should not run outside of a cordova-based project by calling util.isCordova', function(done) {
-            is_cordova.and.returnValue(false);
-            cd_project_root.and.callThrough();  // undo spy here because prepare depends on cdprojectRoot for isCordova check
-            prepare().then(function() {
-                expect('this call').toBe('fail');
-            }, function(err) {
-                expect('' + err).toContain('Current working directory is not a Cordova-based project.');
-            }).fin(done);
+    describe('main method', function () {
+        beforeEach(function () {
+            spyOn(cordova_config, 'read').and.returnValue({});
+            spyOn(restore, 'installPlatformsFromConfigXML').and.returnValue(Q());
+            spyOn(restore, 'installPluginsFromConfigXML').and.returnValue(Q());
+            
+            spyOn(prepare, 'preparePlatforms').and.returnValue(Q);
+        });
+        describe('failure', function () {
+            it('should invoke util.preProcessOptions as preflight task checker, which, if fails, should trigger promise rejection and only fire the before_prepare hook', function(done) {
+                spyOn(util, 'cdProjectRoot').and.returnValue(project_dir);
+                spyOn(util, 'preProcessOptions').and.throwError();
+                prepare({}).fail(function(err) {
+                    expect(err).toBeDefined();
+                    expect(HooksRunner.prototype.fire.calls.count()).toBe(1);
+                    expect(HooksRunner.prototype.fire.calls.argsFor(0)[0]).toEqual('before_prepare');
```

Comment:
I'd prefer we use the recommended jasmine syntax of `toHaveBeenCalled()` if we just want to check whether a spy was called, or `toHaveBeenCalledWith(arg1, arg2)` if we want to set expectations on invocation arguments. Getting deep into the `calls` sub-properties can get ugly and confusing fast.

e.g. `expect(HooksRunner.prototype.fire).toHaveBeenCalledWith('before_prepare', jasmine.any(Object));`